packages/mui-base/src/Tooltip/Popup/useTooltipPopup.ts

```diff
@@ -0,0 +1,319 @@
+    padding: collisionPadding,
+    boundary: collisionBoundary,
+  });
+
+  // https://floating-ui.com/docs/flip#combining-with-shift
+  if (alignment !== 'center') {
+    middleware.push(flipMiddleware, shiftMiddleware);
+  } else {
+    middleware.push(shiftMiddleware, flipMiddleware);
+  }
+
+  middleware.push(
+    size({
+      boundary: collisionBoundary,
+      padding: collisionPadding,
+      apply({ elements: { floating }, rects: { reference }, availableWidth, availableHeight }) {
+        Object.entries({
+          '--available-width': `${availableWidth}px`,
+          '--available-height': `${availableHeight}px`,
+          '--anchor-width': `${reference.width}px`,
+          '--anchor-height': `${reference.height}px`,
+        }).forEach(([key, value]) => {
+          floating.style.setProperty(key, value);
+        });
+      },
+    }),
+    arrow({ element: arrowRef, padding: arrowPadding }),
+    hideWhenDetached && hide(),
+    {
+      name: 'transformOrigin',
+      fn({ elements, middlewareData, placement: renderedPlacement }) {
+        const currentRenderedSide = getSide(renderedPlacement);
+        const arrowEl = arrowRef.current;
+        const arrowX = middlewareData.arrow?.x ?? 0;
+        const arrowY = middlewareData.arrow?.y ?? 0;
+        const arrowWidth = arrowEl?.clientWidth ?? sideOffset;
+        const arrowHeight = arrowEl?.clientHeight ?? sideOffset;
+        const transformX = arrowX + arrowWidth / 2;
+        const transformY = arrowY + arrowHeight;
+
+        const transformOrigin = {
+          top: `${transformX}px calc(100% + ${arrowHeight}px)`,
+          bottom: `${transformX}px ${-arrowHeight}px`,
+          left: `calc(100% + ${arrowHeight}px) ${transformY}px`,
+          right: `${-arrowHeight}px ${transformY}px`,
+        }[currentRenderedSide];
+
+        elements.floating.style.setProperty('--transform-origin', transformOrigin);
+
+        return {};
+      },
+    },
+  );
+
+  const {
+    refs,
+    elements,
+    floatingStyles,
+    context,
+    middlewareData,
+    update,
+    placement: renderedPlacement,
+  } = useFloating({
+    placement,
+    middleware,
+    open,
+    whileElementsMounted: keepMounted ? undefined : autoUpdate,
+    elements: anchor && isElement(anchor) ? { reference: anchor } : undefined,
+    onOpenChange(openValue, eventValue, reasonValue) {
+      onOpenChange(openValue, eventValue, reasonValue);
+
+      const isFocusOpen = openValue && reasonValue === 'focus';
+      const isDismissClose =
+        !openValue && (reasonValue === 'reference-press' || reasonValue === 'escape-key');
+
+      if (isFocusOpen || isDismissClose) {
+        setInstantTypeState(isFocusOpen ? 'focus' : 'dismiss');
+      } else {
+        setInstantTypeState(undefined);
+      }
+
+      const popupElement = refs.floating.current?.firstElementChild;
+      if (!openValue && popupElement) {
+        const computedStyles = ownerWindow(popupElement).getComputedStyle(popupElement);
```

Comment:
How will this work with animation libraries such as React Spring?

Reply:
They can use the `keepMounted` prop on `Tooltip.Popup`. This lets another library control the mounting.

```js
<Tooltip.Root open={open} onOpenChange={onOpenChange}>
  <Tooltip.Trigger>
    <button />
  </Tooltip.Trigger>
  <AnimatePresence>
    {open && (
      <Tooltip.Popup 
        keepMounted 
        render={<motion.div ... />} 
      />
    )}
  </AnimatePresence>
</Tooltip.Root>
```